test/e2e/testsuites/dynamically_provisioned_volume_replicas_across_zones.go

```diff
@@ -0,0 +1,117 @@
+	testtypes "sigs.k8s.io/azuredisk-csi-driver/test/types"
+)
+
+type DynamicallyProvisionedVolumeReplicasAcrossZones struct {
+	Pod                    testtypes.PodDetails
+	CSIDriver              driver.DynamicPVTestDriver
+	Volume                 testtypes.VolumeDetails
+	AzDiskClient           *azDiskClientSet.Clientset
+	StorageClassParameters map[string]string
+}
+
+func (t *DynamicallyProvisionedVolumeReplicasAcrossZones) Run(client clientset.Interface, namespace *v1.Namespace, schedulerName string) {
+	_, maxMountReplicaCount := azureutils.GetMaxSharesAndMaxMountReplicaCount(t.StorageClassParameters, false)
+	// create the storageClass
+	tsc, tscCleanup := t.Pod.Volumes[0].CreateStorageClass(client, namespace, t.CSIDriver, t.StorageClassParameters)
+	defer tscCleanup()
+
+	// create the pod
+	t.Pod.Volumes[0].StorageClass = tsc.StorageClass
+	tpod, cleanups := t.Pod.SetupWithDynamicVolumes(client, namespace, t.CSIDriver, t.StorageClassParameters, schedulerName)
+	for i := range cleanups {
+		defer cleanups[i]()
+	}
+
+	ginkgo.By("deploying the pod")
+	tpod.Create()
+	defer tpod.Cleanup()
+	ginkgo.By("checking that the pod is running")
+	tpod.WaitForRunning()
+
+	pod := tpod.Pod.DeepCopy()
+	volume := pod.Spec.Volumes[0]
+
+	ctx := context.Background()
+
+	if volume.PersistentVolumeClaim == nil {
+		framework.Failf("volume (%s) does not hold PersistentVolumeClaim field", volume.Name)
+	}
+	pvc, err := client.CoreV1().PersistentVolumeClaims(tpod.Namespace.Name).Get(ctx, volume.PersistentVolumeClaim.ClaimName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+
+	pv, err := client.CoreV1().PersistentVolumes().Get(ctx, pvc.Spec.VolumeName, metav1.GetOptions{})
+	framework.ExpectNoError(err)
+	framework.ExpectNotEqual(pv.Spec.CSI, nil)
+	framework.ExpectEqual(pv.Spec.CSI.Driver, consts.DefaultDriverName)
+
+	diskName, err := azureutils.GetDiskName(pv.Spec.CSI.VolumeHandle)
+	framework.ExpectNoError(err)
+
+	// Confirm that the primary and the replica AzVolumeAttachments are created on different zones
+	err = wait.PollImmediate(testconsts.Poll, testconsts.PollTimeout,
+		func() (bool, error) {
+
+			labelSelector := labels.NewSelector()
+			volReq, err := azureutils.CreateLabelRequirements(consts.VolumeNameLabel, selection.Equals, diskName)
+			framework.ExpectNoError(err)
+			labelSelector = labelSelector.Add(*volReq)
+
+			azVolumeAttachments, err := t.AzDiskClient.DiskV1alpha2().AzVolumeAttachments(consts.DefaultAzureDiskCrdNamespace).List(ctx, metav1.ListOptions{LabelSelector: labelSelector.String()})
+			if err != nil {
+				return false, err
+			}
+
+			if azVolumeAttachments == nil {
+				return false, nil
+			}
+
+			klog.Infof("found %d AzVolumeAttachments for volume (%s)", len(azVolumeAttachments.Items), diskName)
+
+			if len(azVolumeAttachments.Items) != maxMountReplicaCount+1 {
+				return false, nil
+			}
+
+			nodeObj1, err := client.CoreV1().Nodes().Get(ctx, azVolumeAttachments.Items[0].Spec.NodeName, metav1.GetOptions{})
```

Comment:
Rather than assume two replicas, consider using a `for` loop to get each attachment's node and check for duplicates through a `map[string]struct{}` where the key is the zone name?

Reply:
Yeah, that is cleaner but then we'll have to find the the number of availability zones too. Say, if replicas=3 and number of
zones is 2, 2 of the replicas will be on the same zone. We'll have to verify that the count of replicas per zone <= replicas/(num of zones) + 1. Will see if there's a way to get the number of zones of the cluster from the e2e tests.